text/0118-priming-hints.md

```diff
@@ -0,0 +1,175 @@
+
+These comments would be used on property getters, and their format would be:
+
+```js
+/**
+ * priming-hint: <scope> <value>
+ */
+```
+
+With the following values for "scope":
+- always: priming mechanisms should always evaluate the given property with the given value for the purposes of priming through the component tree
+- offline: priming mechanisms should evaluate the given property with the given value for the purposes of priming through the component tree only if the priming is targeted at an offline use case
+
+We differentiate between these 2 scopes because online and offline priming use cases can be qualitatively different. Online priming is aimed at enhancing performance of rendering, with the assumption that further resources and data can be pulled from the server later on as necessary. Offline priming aims to prime a complete useful set of resources and data from the server up-front with no additional requests later. In general, online priming is more sensitive to performance and offline priming is more sensitive to missing scope.
+
+For our bar example, we could modify the barItem component by replacing the use of a `@track` member with a new `@api` property that uses could make use of this priming hint to guide a priming mechanism to always prime its contents:
+
+```js
+export default class BarItem extends LightningElement {
+
+    /**
+     * priming-hint: always true
+     */
+    @api
+    viewContent() {
+        return this._loadContent;
+    }
+
+    @api
+    loadContent() {
+        this._loadContent = true;
+
+        this.dispatchEvent(new CustomEvent('active'));
+    }
+
+    ...
+}
+```
+
+```html
+<!-- barItem.html -->
+<template>
+    <template if:true={viewContent}>
+        <slot></slot>
+    </template>
+</template>
+```
+
+## Other Possible Strategies
+
+Here is a brief survey of other strategies we considered for priming hints.
+
+### Contextual Import
+
+```js
+import {primingContext} from `@salesforce/priming`;
+...
+get loadContent {
+    return primingContext?.isPriming || _loadContent;
+}
+```
+
+Here we have the property itself explicitly import an indicator of priming status and use it to drive the value to return for the property. This feels like it would be too much work for developers and they could easily calculate the property incorrectly.
+
+### Annotation
+
+```js
+import {forcePriming} from `@salesforce/priming`;
+...
+
+@forcePriming
+get loadContent {
+    return _loadContent;
+}
+```
+
+Here the component opts in to the priming-specific override by using a decorator. Using decorators like this seems out-of-step with other parts of the LWC specification.
+
+### Metadata XML
+
+```xml
+<?xml version="1.0" encoding="UTF-8"?>
+<LightningComponentBundle xmlns="xmlns=http://soap.sforce.com/2006/04/metadata">
+    <isExposed>true</isExposed>
+    <primingHints>
+        <primingHint value="true">loadContent</primingHint>
+    </primingHints>
+</LightningComponentBundle>
+```
+
```

Comment:
> @shesmer The question for me here is whether we ever want to elevate these priming hints in the LWC programming model or whether it says an internal core only technology...
> Using this xml would work well for internal core only but not allow us to expose this any other way.

Another drawback of this programming model is that it separates out the primingHint metadata away from the code. These makes it more likely the developer would, say, forget to update the XML when they rename its corresponding function.

We're advocating that this be a public feature, but we can discuss if that's really the best course. @pmdartus I know had some reservations on making this public.

Reply:
> @pmdartus I know had some reservations on making this public.

Here are my 2 cents on where we should have this discussion. 

The LWC team uses this repo the `lwc-rfcs` repo track proposals related to the open-source LWC framework. Salesforce-only / internal proposals are usually tracked internally using Quip documents. The repository is public, and I don’t think it is the correct place to discuss internal details. 

Regardless of the approach, we take (a comment, field in meta XML, or a decorator), this information will not be consumed by the LWC framework at runtime. The LWC engine isn't, and shouldn't be, aware of how data is retrieved. It's an implementation detail from its standpoint. 

I would recommend moving this proposal somewhere else.

Reply:
@pmdartus @shesmer @gonzalocordero I have imported this RFC into an internal quip doc, please look for your mentions.